steering_controllers_library/test/test_steering_controllers_library.cpp

```diff
@@ -0,0 +1,200 @@
+#include <utility>
+#include <vector>
+
+#include "hardware_interface/types/hardware_interface_type_values.hpp"
+
+class SteeringControllersLibraryTest
+: public SteeringControllersLibraryFixture<TestableSteeringControllersLibrary>
+{
+};
+
+// checking if all interfaces, command, state and reference are exported as expected
+TEST_F(SteeringControllersLibraryTest, check_exported_intefaces)
+{
+  SetUpController();
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  auto command_intefaces = controller_->command_interface_configuration();
+  ASSERT_EQ(command_intefaces.names.size(), joint_command_values_.size());
+  EXPECT_EQ(
+    command_intefaces.names[CMD_TRACTION_RIGHT_WHEEL],
+    rear_wheels_names_[0] + "/" + traction_interface_name_);
+  EXPECT_EQ(
+    command_intefaces.names[CMD_TRACTION_LEFT_WHEEL],
+    rear_wheels_names_[1] + "/" + traction_interface_name_);
+  EXPECT_EQ(
+    command_intefaces.names[CMD_STEER_RIGHT_WHEEL],
+    front_wheels_names_[0] + "/" + steering_interface_name_);
+  EXPECT_EQ(
+    command_intefaces.names[CMD_STEER_LEFT_WHEEL],
+    front_wheels_names_[1] + "/" + steering_interface_name_);
+
+  auto state_intefaces = controller_->state_interface_configuration();
+  ASSERT_EQ(state_intefaces.names.size(), joint_state_values_.size());
+  EXPECT_EQ(
+    state_intefaces.names[STATE_TRACTION_RIGHT_WHEEL],
+    controller_->rear_wheels_state_names_[0] + "/" + traction_interface_name_);
+  EXPECT_EQ(
+    state_intefaces.names[STATE_TRACTION_LEFT_WHEEL],
+    controller_->rear_wheels_state_names_[1] + "/" + traction_interface_name_);
+  EXPECT_EQ(
+    state_intefaces.names[STATE_STEER_RIGHT_WHEEL],
+    controller_->front_wheels_state_names_[0] + "/" + steering_interface_name_);
+  EXPECT_EQ(
+    state_intefaces.names[STATE_STEER_LEFT_WHEEL],
+    controller_->front_wheels_state_names_[1] + "/" + steering_interface_name_);
+
+  // check ref itfs
+  auto reference_interfaces = controller_->export_reference_interfaces();
+  ASSERT_EQ(reference_interfaces.size(), joint_reference_interfaces_.size());
+  for (size_t i = 0; i < joint_reference_interfaces_.size(); ++i)
+  {
+    const std::string ref_itf_name =
+      std::string(controller_->get_node()->get_name()) + "/" + joint_reference_interfaces_[i];
+    EXPECT_EQ(reference_interfaces[i].get_name(), ref_itf_name);
+    EXPECT_EQ(reference_interfaces[i].get_prefix_name(), controller_->get_node()->get_name());
+    EXPECT_EQ(reference_interfaces[i].get_interface_name(), joint_reference_interfaces_[i]);
+  }
+}
+
+// Tests controller update_reference_from_subscribers and
+// its two cases for position_feedback true/false behavior
+// when too old msg is sent i.e age_of_last_command > ref_timeout case
+TEST_F(SteeringControllersLibraryTest, test_both_update_methods_for_ref_timeout)
+{
+  SetUpController();
+
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
```

Comment:
These test implementations tend to be pretty flaky. Why do you need an entire node environment for testing the business logic? Simple unit tests for *a library* are much more powerful IMO. I'd do the brunt of the testing against steering_odometry in pure C++

Reply:
yes, we need this since we are loading all parameters through the node interface each time. What is the value in implementing fake things if we anyway have that things are working with provided parameters.